File: Python/binary_search.py
# Binary Search
# Works for sorted arrays
# Tested with Python 3.13.2

def binary_search(arr, value):
    right = len(arr) - 1 # Last index
    left = 0 # First index
    
    while left <= right:
        mid = (left + right) // 2 # Get the mid value

        if arr[mid] == value:
            return mid
        
        if arr[mid] < value: # Value is larger
            left = mid + 1
        else: # Value is smaller
            right = mid - 1

        # When using `left = mid` and `right = mid` without `+1` and `-1`,
        # it would cause an infinity loop when value is last or
        # first.
        
    return -1

# USAGE

my_array  = [2, 3, 7, 7, 11, 15, 25, 28]
i = binary_search(my_array, 15)
print(i)

File: CPP/binarySearch.cpp
#include <iostream>

using namespace std;

template <size_t N>
int binarySearch(int (&arr)[N], int value);

int main() {
    int myArray[] = {2, 3, 7, 7, 11, 15, 25};
    
    int i = binarySearch(myArray, 25);
    cout << i << endl;

    return 0;
}

template <size_t N>
int binarySearch(int (&arr)[N], int value) {
    int left = 0; // Least index
    int right = N - 1; // Max index
    int mid; // Index of the mid element

    while (left <= right) {
        mid = (left + right) / 2;
        int midValue = arr[mid];

        if (midValue == value) {
            return mid;
        }
        else if (midValue < value) {
            left = mid + 1;
        }
        else {
            right = mid - 1;
        }
    }

    return - 1; // Value not found
}